R/plotly_build.R

```diff
@@ -0,0 +1,373 @@
+#' 'Build' (i.e., evaluate) a plotly object
+#' 
+#' This generic function creates the list object sent to plotly.js
+#' for rendering. Using this function can be useful for overriding defaults
+#' provided by \code{ggplotly}/\code{plot_ly} or for debugging rendering
+#' errors.
+#' 
+#' @param p a ggplot object, or a plotly object, or a list.
+#' @export
+#' @examples
+#' 
+#' p <- plot_ly(economics, x = ~date, y = ~pce)
+#' # the unevaluated plotly object
+#' str(p)
+#' # the evaluated data
+#' str(plotly_build(p)$x$data)
+#' 
+plotly_build <- function(p) {
+  UseMethod("plotly_build")
+}
+
+#' @export
+plotly_build.list <- function(p) {
+  as.widget(p)
+}
+
+#' @export
+plotly_build.gg <- function(p) {
+  ggplotly(p)
+}
+
+#' @export
+plotly_build.plotly <- function(p) {
+  
+  layouts <- Map(function(x, y) {
+    
+    d <- plotly_data(p, y)
+    
+    x <- rapply(x, eval_attr, data = d, how = "list")
+    x[lengths(x) > 0]
+    
+  }, p$x$layoutAttrs, names(p$x$layoutAttrs))
+  
+  # get rid of the data -> layout mapping and merge all the layouts
+  # into a single layout (more recent layouts will override older ones)
+  p$x$layoutAttrs <- NULL
+  p$x$layout <- modify_list(p$x$layout, Reduce(modify_list, layouts))
+  
+  # If type was not specified in plot_ly(), it doesn't create a trace unless
```

Comment:
If I comment out, then both the smoother and raw line are plotted, so this does seem to explain what is happening, but I don't understand why.